File: backend/app/continuous_types.py
"""
Canonical string-literal vocabulary for Phase 7 continuous sessions — the
backend's single source of truth for every enum-like string used across
continuous_models.py, continuous_schemas.py, continuous_routes.py, and
risk.py.

The frontend mirrors these exact values as JSDoc typedefs in
frontend/src/realityCheck/continuousTypes.js. If a value here changes, it
must change there too, in the same commit — that file's header comment
points back here. There is no shared package between the two runtimes
(plain JS, not TS), so this pairing is enforced by convention plus the
cross-referencing comments, not by the type system.
"""

SESSION_STATES = ("CREATED", "ACTIVE", "ENDED", "CANCELLED", "ERROR")
TERMINAL_SESSION_STATES = ("ENDED", "CANCELLED", "ERROR")

EVENT_TYPES = (
    "session_started", "session_ended", "session_cancelled", "session_error",
    "monitoring_started", "monitoring_stopped",
    "face_missing", "face_restored", "multiple_faces_detected", "multiple_faces_cleared",
    "landmark_discontinuity", "frozen_frame_suspected",
    "camera_track_ended", "camera_stream_interrupted", "camera_device_changed",
    "challenge_requested", "challenge_started", "challenge_passed",
    "challenge_failed", "challenge_aborted", "challenge_timeout",
    "risk_state_changed", "technical_error",
)

SEVERITIES = ("info", "warning", "suspicious", "error")

# Head Turn types reuse the frontend's existing challengeConstants.js
# CHALLENGE_TYPES vocabulary verbatim (TURN_HEAD_LEFT / TURN_HEAD_RIGHT).
# Phase 7 does not introduce a new challenge type — see ground rules.
# Phase 10 adds DEPTH_PROXIMITY ("move closer to the camera") as the third
# challenge type. Scheduling (request_next_challenge's secrets.choice) and
# risk scoring (risk.py) are both already type-agnostic, so no other backend
# change is required to add a challenge type beyond this tuple — see
# frontend/src/constants/depthProximityConstants.js for its config/state
# machine and frontend/src/realityCheck/continuousTypes.js for the mirror.
CHALLENGE_TYPES = ("TURN_HEAD_LEFT", "TURN_HEAD_RIGHT", "LIGHT", "DEPTH_PROXIMITY")

CHALLENGE_TRIGGERS = ("RANDOM", "EVENT")

CHALLENGE_STATUSES = ("PENDING", "PASSED", "FAILED", "TIMEOUT", "ABORTED")
# Subset the client is allowed to report as a result — PENDING is server-only
# (it's the state a freshly-issued challenge starts in).
CHALLENGE_CLIENT_OUTCOMES = ("PASSED", "FAILED", "TIMEOUT", "ABORTED")

RISK_STATES = ("LOW_RISK", "REVIEW_RECOMMENDED", "INCONCLUSIVE")

# Phase 11 — engineering/research-only ground-truth labels for calibration
# datasets (see app/continuous_models.py's set_ground_truth_label and
# app/fusion.py's module docstring). Never surfaced to or settable by the
# candidate-facing production verification flow; a dev/offline tool attaches
# these to already-ended sessions purely for later score-distribution
# comparison. GENUINE is the only "expected good" label; the rest are
# labeled attack categories a calibration dataset might record.
GROUND_TRUTH_LABELS = ("GENUINE", "PRINT_ATTACK", "PHONE_REPLAY", "FACE_SWAP", "MASK", "OTHER")

File: frontend/src/realityCheck/continuousTypes.js
/**
 * Canonical string-literal vocabulary for Phase 7 continuous sessions —
 * the frontend's single reference point for every enum-like string used
 * across the passive monitor, challenge scheduler, and the public
 * `createRealityCheckSession` interface (Stage 3/4).
 *
 * This project is plain JS, not TypeScript, so there is no shared type
 * package between frontend and backend. The canonical values live in
 * backend/app/continuous_types.py — that Python module is the source of
 * truth (it's what actually validates every request via Pydantic). This
 * file is the frontend mirror: if a value changes in continuous_types.py,
 * it must change here too, in the same commit, or the frontend will start
 * sending strings the backend's field_validators reject with 422.
 */

/**
 * @typedef {'CREATED'|'ACTIVE'|'ENDED'|'CANCELLED'|'ERROR'} SessionState
 * Continuous-session lifecycle. ENDED, CANCELLED, and ERROR are terminal
 * and irreversible — mirrors backend/app/continuous_models.py's
 * TERMINAL_SESSION_STATES.
 */
export const SESSION_STATES = /** @type {const} */ ([
  'CREATED', 'ACTIVE', 'ENDED', 'CANCELLED', 'ERROR'
]);
export const TERMINAL_SESSION_STATES = /** @type {const} */ (['ENDED', 'CANCELLED', 'ERROR']);

/**
 * @typedef {(
 *   'session_started'|'session_ended'|'session_cancelled'|'session_error'|
 *   'monitoring_started'|'monitoring_stopped'|
 *   'face_missing'|'face_restored'|'multiple_faces_detected'|'multiple_faces_cleared'|
 *   'landmark_discontinuity'|'frozen_frame_suspected'|
 *   'camera_track_ended'|'camera_stream_interrupted'|'camera_device_changed'|
 *   'challenge_requested'|'challenge_started'|'challenge_passed'|
 *   'challenge_failed'|'challenge_aborted'|'challenge_timeout'|
 *   'risk_state_changed'|'technical_error'
 * )} EventType
 * Fixed vocabulary — see Phase 7 spec §5. Must exactly match
 * backend/app/continuous_types.py's EVENT_TYPES tuple (validated there via
 * a Pydantic field_validator; an unrecognized value is rejected with 422).
 */
export const EVENT_TYPES = /** @type {const} */ ([
  'session_started', 'session_ended', 'session_cancelled', 'session_error',
  'monitoring_started', 'monitoring_stopped',
  'face_missing', 'face_restored', 'multiple_faces_detected', 'multiple_faces_cleared',
  'landmark_discontinuity', 'frozen_frame_suspected',
  'camera_track_ended', 'camera_stream_interrupted', 'camera_device_changed',
  'challenge_requested', 'challenge_started', 'challenge_passed',
  'challenge_failed', 'challenge_aborted', 'challenge_timeout',
  'risk_state_changed', 'technical_error'
]);

/**
 * @typedef {'info'|'warning'|'suspicious'|'error'} Severity
 */
export const SEVERITIES = /** @type {const} */ (['info', 'warning', 'suspicious', 'error']);

/**
 * @typedef {'TURN_HEAD_LEFT'|'TURN_HEAD_RIGHT'|'LIGHT'|'DEPTH_PROXIMITY'} ChallengeType
 * TURN_HEAD_LEFT/TURN_HEAD_RIGHT reuse the existing
 * constants/challengeConstants.js CHALLENGE_TYPES values verbatim — Phase 7
 * does not introduce a new challenge type. DEPTH_PROXIMITY ("move closer to
 * the camera") is added in Phase 10 — see
 * constants/depthProximityConstants.js.
 */
export const CHALLENGE_TYPES = /** @type {const} */ (['TURN_HEAD_LEFT', 'TURN_HEAD_RIGHT', 'LIGHT', 'DEPTH_PROXIMITY']);

/**
 * @typedef {'RANDOM'|'EVENT'} ChallengeTrigger
 */
export const CHALLENGE_TRIGGERS = /** @type {const} */ (['RANDOM', 'EVENT']);

/**
 * @typedef {'PENDING'|'PASSED'|'FAILED'|'TIMEOUT'|'ABORTED'} ChallengeStatus
 * PENDING is server-only (the state a freshly-issued challenge starts in);
 * the client only ever reports one of CHALLENGE_CLIENT_OUTCOMES.
 */
export const CHALLENGE_STATUSES = /** @type {const} */ (['PENDING', 'PASSED', 'FAILED', 'TIMEOUT', 'ABORTED']);

/** @typedef {'PASSED'|'FAILED'|'TIMEOUT'|'ABORTED'} ChallengeOutcome */
export const CHALLENGE_CLIENT_OUTCOMES = /** @type {const} */ (['PASSED', 'FAILED', 'TIMEOUT', 'ABORTED']);

/**
 * @typedef {'LOW_RISK'|'REVIEW_RECOMMENDED'|'INCONCLUSIVE'} RiskState
 * INCONCLUSIVE means "we cannot say", not "high risk" — see
 * backend/app/risk.py's module docstring. UI copy must not treat it as a
 * risk tier.
 */
export const RISK_STATES = /** @type {const} */ (['LOW_RISK', 'REVIEW_RECOMMENDED', 'INCONCLUSIVE']);

/**
 * @typedef {Object} ChallengeResultSubmission
 * The exact shape POSTed to
 * `/sessions/continuous/{sessionId}/challenges/{challengeId}/result`
 * (see backend/app/continuous_schemas.py's ChallengeResultRequest).
 * @property {string} nonce - Must match the nonce issued with this challenge.
 * @property {ChallengeOutcome} outcome
 * @property {Object|null} [detail] - Free-form diagnostic detail (e.g. challenge telemetry summary).
 */

/**
 * @typedef {Object} NextChallenge
 * The shape returned by `GET /sessions/continuous/{sessionId}/next-challenge`
 * when a challenge is issued.
 * @property {string} challengeId
 * @property {ChallengeType} type
 * @property {string} nonce
 * @property {string} expiresAt - ISO 8601 UTC.
 */

/**
 * @typedef {Object} NoNextChallenge
 * The shape returned when no challenge is issued yet.
 * @property {true} none
 * @property {number|null} retryAfterMs
 * @property {string} reason - e.g. "COOLDOWN", "CHALLENGE_IN_FLIGHT", "MAX_CHALLENGES_REACHED".
 */
